test/distributed/org/apache/cassandra/distributed/impl/Instance.java

```diff
@@ -651,7 +666,28 @@ public void startup(ICluster cluster)
                     throw (RuntimeException) t;
                 throw new RuntimeException(t);
             }
-        }).run();
+        }).call();
+        DurationSpec timeout = startupTimeout();
+        if (timeout == null)
+        {
+            waitOn(result);
+        }
+        else
+        {
+            try
+            {
+                result.get(timeout.quantity(), timeout.unit());
+            }
+            catch (InterruptedException e)
+            {
+                Thread.currentThread().interrupt();
```

Comment:
Could you clarify? I thought if we catch interrupted exception, thread is already interrupted?

Reply:
there are 2 methods on `Thread` and it depends on what gets called

`interrupted()` this clears the result as you are expected to handle
`isInterrupted()` this does not clear

You as the catcher have no clue what was called!  In Effective Java the recommended pattern is to re-interrupt the thread (it's just a flag) to make sure that the caller that catches your throwable can still see the interrupt.

If you do not re-interrupt the thread then its possible that this gets ignored and we swallow the interrupt; which is why effective java tells you to do this.

There are cases where clearing it makes sense, such as being interrupted to *release resources*, but since our code base doesn't do that, it only means "shutdown" for us, so making sure we don't loose the interrupt is best as general statement

Reply:
For example `java.util.concurrent.CompletableFuture#timedGet` does

```
if (Thread.interrupted())
  return null; (reportGet will then do the throw new InterruptedException();)
```

This means the flag gets cleared!  So if we catch and and don't release the thread, then its possible that the interrupt gets lost and the thread stays up... so we should always re-interrupt

Reply:
Ah nice, good to know, thank you for elaborating!